python/cudf/cudf/options.py

```diff
@@ -0,0 +1,92 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.
+
+from dataclasses import dataclass
+from typing import Any, Callable, Dict, Optional, Union
+
+
+@dataclass
+class CUDFOption:
+    value: Any
+    description: str
+    validator: Callable
+
+
+_CUDF_OPTIONS: Dict[str, CUDFOption] = {}
+
+
+def _register_option(
+    name: str, default_value: Any, description: str, validator: Callable
+):
+    """Register an entry in the option dictionary.
+
+    Parameters
+    ----------
+    name : str
+        The name of the option. Also used as the key in the dictionary.
+
+    default_value : Any
+        The default value of the option.
+
+    description : str
+        A text description of the option.
+
+    validator : Callable
+        A function that returns ``True`` if a given value is valid for the
+        option, ``False`` otherwise.
+    """
+    if not validator(default_value):
+        raise ValueError(f"Invalid default value: {default_value}")
+
+    _CUDF_OPTIONS[name] = CUDFOption(default_value, description, validator)
+
+
+def get_option(key: str) -> Any:
+    """Get the value of option.
+
+    Parameters
+    ----------
+    key : str
+        The name of the option.
+
+    Returns
+    -------
+    The value of the option.
+    """
+    return _CUDF_OPTIONS[key].value
+
+
+def set_option(key: str, val: Any):
+    """Set the value of option.
+
+    Raises ``ValueError`` if the provided value is invalid.
+
+    Parameters
+    ----------
+    key : str
+        The name of the option.
+    val : Any
+        The value to set.
+    """
+    config = _CUDF_OPTIONS[key]
+    if not config.validator(val):
+        raise ValueError(f"Invalid option {val}")
```

Comment:
pandas leaves it to the validator to raise. I think that's a better approach because of the flexibility it offers.

https://github.com/pandas-dev/pandas/blob/87930ef48c7fd3090f914eb5020b5371943a2216/pandas/_config/config.py#L469-L471